compiler/src/org.graalvm.compiler.phases.common/src/org/graalvm/compiler/phases/common/LoweringPhase.java

```diff
@@ -236,14 +238,25 @@ private void setLastFixedNode(FixedWithNextNode n) {
             assert n.isAlive() : n;
             lastFixedNode = n;
         }
+
+        @Override
+        public Boolean getExpandsVolatileMembarsLate() {
+            return expandsVolatileMembarsLate;
+        }
     }
 
     private final CanonicalizerPhase canonicalizer;
     private final LoweringTool.LoweringStage loweringStage;
+    private Boolean expandsVolatileMembarsLate;
 
-    public LoweringPhase(CanonicalizerPhase canonicalizer, LoweringTool.LoweringStage loweringStage) {
+    public LoweringPhase(CanonicalizerPhase canonicalizer, LoweringTool.LoweringStage loweringStage, Boolean expandsVolatileMembarsLate) {
```

Comment:
Why is this a boxed Boolean? Also, I only see it ever being set to true - under what conditions do you expect it to be set to false?

Reply:
Thanks for looking at this.
The boxed Boolean is a mistake (must have trusted the IDE refactoring too much).
There are 2 LoweringPhase constructors. Second one doesn't take the expandsVolatileMembarsLate argument and sets it to false.That one is called from the EconomyHighTier constructor. So expandsVolatileMembarsLate should only be true when called from the HighTier constructor.
This said I now wonder if that does the right thing: snippets would always have expandsVolatileMembarsLate set to false and I suppose a snippet could have a volatile access too.
An other way of doing that would be to force LateMembars to false in the OptionValues in the case of the economic configuration.

Reply:
Sorry, I missed the fact that you added an extra LoweringPhase constructor.
@tkrodriguez can you please weigh in on Roland's observation about volatiles in snippets.

Reply:
Some context: the end goal with all this is to generate a code pattern on aarch64 that doesn't contain membar instructions. Using that code pattern is only legal if all code that is emitted uses the membar less pattern. Mixing membar volatile accesses with non membar volatile accesses in the same process must not happen. So:
on non aarch64, we want LateMembars true for the CommunityCompilerconfiguration and false for the EconomyCompilerConfiguration
on aarch64 once the non membar pattern is supported, LateMembars would need to be true for both the CommunityCompilerconfiguration and EconomyCompilerConfiguration.
The patch I proposed tries to achieve this by overriding the value of LateMembars with an extra boolean that's passed around but it seems hard to do globally. Actually the root cause is that option values are inconsistent. If the economy configuration is enabled then LateMembars must be off. It's a common pattern in hotspot to force values of command line flags so they are consistent. I don't see much of that in Graal.

Reply:
>  If the economy configuration is enabled then LateMembars must be off

Isn't this a matter of ensuring in the `EconomyLowTier` constructor that `LateMembars` is not explicitly set?

It is indeed a little harder to ensure consistency of options in Graal because they are not as global as HotSpot options. This allows for options to be overridden in certain contexts (e.g. testing, snippets) while not disturbing other concurrently executing contexts.

Reply:
> Isn't this a matter of ensuring in the `EconomyLowTier` constructor that `LateMembars` is not explicitly set?

So you're saying changing the value of the LateMembars option is the way to go?

LateMembars is on by default and we want it off at lowering time for the Economy configuration. Once https://github.com/oracle/graal/pull/1772 is in, we would want if off only on x86.

Reply:
Why is `LateMembars` even an option? It sounds like it's controlling something that should never be controlled by a user.

Reply:
> Why is `LateMembars` even an option? It sounds like it's controlling something that should never be controlled by a user.

To help troubleshooting. But you're right that it's not something that the user should change. So if we remove the option where would we store a global flag that tells whether membars should be added at lowering time or as an independent phase or not at all?

Reply:
Do we really need a global? At every point that `LateMembar` is currently read, do we not have sufficient context (i.e. architecture, compiler config, etc) to determine how to treat membars? If not, I would try and thread that info through somehow.

Reply:
From the DefaultJavaLoweringProvider, AFAICT, there's no way to tell what compiler configuration is being used. Threading that info through is what the proposed patch does, right? Except I now think snippets are not covered and they should be. But getting than bit of information through to the snippet machinery seems quite involved. Couldn't it be attached to some objects that's carried around? The Graph object maybe?

Reply:
Where in the snippet machinery is this information needed?

Reply:
SnippetTemplate() creates a LoweringPhase() which AFAIU would also need that information

Reply:
Seems like we want something similar to `org.graalvm.compiler.nodes.StructuredGraph.guardsStage`. What do you think @tkrodriguez @gilles-duboscq ?

Reply:
FYI, I pushed a new change that propagates expandsVolatileMembarsLate to snippets. Wasn't as complicated as I thought but that looks pretty ugly and fragile to me.

Reply:
Doesn't look any worse than threading `loweringStage` through to the snippets in my opinion.